cms/core/tests/__init__.py

```diff
@@ -0,0 +1,67 @@
+import json
+from contextlib import ExitStack, contextmanager
+
+from django.db import connections
+from django.test import TransactionTestCase as _TransactionTestCase
+from django.test.utils import CaptureQueriesContext
+from django.utils.functional import partition
+
+from cms.core.db_router import READ_REPLICA_DB_ALIAS
+
+
+class TransactionTestCase(_TransactionTestCase):
+    """A modified TransactionTestCase which ensures models created during migrations are accessible,
+    and adds helpers for testing with multiple database connections.
+    """
+
+    databases = "__all__"
+
+    # The site depends on instances created during migrations, so rollback must be serialized.
+    serialized_rollback = True
+
+    def _fixture_setup(self):
+        """Set up fixtures for test cases."""
+        if self.serialized_rollback and (
+            fixtures := getattr(connections["default"], "_test_serialized_contents", None)
+        ):
+            # Parse the fixture directly rather than serializing it into models for performance reasons.
+            fixtures = json.loads(fixtures)
+
+            # HACK: Wagtail's locales are read from a pre-save hook, which means they need to exist
+            # very early when loading data.
+            # TODO: Remove after https://github.com/wagtail/wagtail/pull/12649 merges.
+            fixtures, eager_fixtures = partition(lambda item: item["model"] == "wagtailcore.locale", fixtures)
+
+            connections["default"]._test_serialized_contents = json.dumps(eager_fixtures + fixtures)  # pylint: disable=protected-access
+
+        super()._fixture_setup()
+
+    @contextmanager
+    def assertTotalNumQueries(self, num):  # pylint: disable=invalid-name
+        """Assert the total number of queries, regardless of the connection used."""
+        with ExitStack() as context_stack:
+            capture_query_contexts = [
+                context_stack.enter_context(CaptureQueriesContext(conn)) for conn in connections.all()
+            ]
+
+            yield
+
+        queries = sorted(
+            query["sql"] for captured_query_context in capture_query_contexts for query in captured_query_context
+        )
+
+        self.assertEqual(
+            len(queries),
+            num,
+            f"{len(queries)} queries executed, {num} expected\nCaptured queries were (in alphabetical order):\n"
+            + "\n".join(queries),
+        )
+
+    @contextmanager
+    def assertNumQueriesConnection(self, *, default=0, replica=0):  # pylint: disable=invalid-name
+        """Assert the number of queries per connection."""
+        with (
+            self.assertNumQueries(default, using="default"),
```

Comment:
nitpicky suggestion: now that we use `READ_REPLICA_DB_ALIAS`, it may be nice to use `DEFAULT_DB_ALIAS` instead of "default" in tests